server/chat-plugins/datasearch.js

```diff
@@ -1092,36 +1082,58 @@ function runMovesearch(target, cmd, canAll, message) {
 		return {reply: "No search parameters other than 'all' were found. Try '/help movesearch' for more information on this command."};
 	}
 
-	let dex = {};
-	if (targetMons.length) {
-		let intersectionOfLearnsets = [];
-		for (let mon in lsetData) {
-			if (intersectionOfLearnsets.length) {
-				intersectionOfLearnsets = intersectionOfLearnsets.filter(move => {
-					return lsetData[mon][move] !== undefined;
-				});
-			} else {
-				intersectionOfLearnsets = Object.keys(lsetData[mon]);
-			}
+	const getFullLearnsetOfPokemon = (template) => {
+		if (!template.learnset) template = Dex.getTemplate(template.baseSpecies);
+		let lsetData = new Set();
+		for (const move in template.learnset) {
+			lsetData.add(move);
 		}
 
-		let finalizedLearnset = {};
-		for (let move of intersectionOfLearnsets) {
-			finalizedLearnset[move] = Object.values(lsetData)[0][move];
+		while (template.prevo) {
+			template = Dex.getTemplate(template.prevo);
+			for (const move in template.learnset) {
+				lsetData.add(move);
+			}
 		}
 
-		for (let move in finalizedLearnset) {
-			dex[move] = Dex.getMove(move);
-		}
-	} else {
-		for (let move in Dex.data.Movedex) {
-			dex[move] = Dex.getMove(move);
+		return lsetData;
+	};
+
+	// Since we assume we have no target mons at first
+	// then the valid moveset we can search is the set of all moves.
+	let validMoves = new Set();
```

Comment:
So, I probably wasn't clear, I apologize - I shouldn't have cited "implicit array creation". 

For initialization (like here and elsewhere below), something like `new Set(Object.keys(Dex.data.Movedex))` should be both more clear and pretty much as performant (yes, you create an array for `Object.keys`, but `Set` can theoretically use the `.length` hit to prealloc space better. Its also still `N` ops performance wise).

My complaint is with using the `set = new Set([...set, ...adds])` pattern for addition, especially when `set` is already large - you're re-adding all of `set`, making the operation `N + M` (which `N` = set.size and `M` = adds.size) when it should just be `M`.